proxy/envoy/config.go

```diff
@@ -664,15 +664,22 @@ func buildEgressFromSidecarVirtualHostOnPort(rule *proxyconfig.EgressRule, mesh
 	externalTrafficRoute := buildDefaultRoute(externalTrafficCluster)
 
 	return &VirtualHost{
-		Name:    rule.Domains[0] + "-" + strconv.Itoa(port.Port),
+		Name:    ruleKey + ":" + strconv.Itoa(port.Port),
 		Domains: appendPortToDomains(rule.Domains, port.Port),
 		Routes:  []*HTTPRoute{externalTrafficRoute},
 	}
 }
 
 func buildEgressFromSidecarHTTPRoutes(mesh *proxyconfig.ProxyMeshConfig, egressRules map[string]*proxyconfig.EgressRule,
 	httpConfigs HTTPRouteConfigs) HTTPRouteConfigs {
-	for _, rule := range egressRules {
+
+	egressRules, errs := model.RejectConflictingEgressRules(egressRules)
+
+	if errs != nil {
+		glog.Warningf("Rejected rules: %v", errs)
+	}
```

Comment:
I don't think we should be doing validation in the config generation phase. Please take a look at validation.go. We validate when users submit rules, so that we can provide immediate feedback. All rule validations need to be done before we enter config generation.

Reply:
This is the problem - the validation is performed on a single configuration object only. This way you cannot find conflicts between different configuration objects. 

I propose the following - today perform this check in the configuration generation check. Later consider implementing validation of conflicts of all Istio configuration objects, probably on the server-side, using web-hooks. The check during the configuration generation could still be relevant, as a precaution, in case any bugs happen between updating configuration objects and reading them. 

CC @kyessenov .

Reply:
I see. In that case, I think we should normalize the rules (merge) and not reject. This is because, we do not have any actions on the egress rules. Its just a list of domains to pass through from the sidecar. In case there are two egress rules with different settings for use_egress_proxy, just take the first one and ignore the rest. This is the mode we adopt throughout this code. I do not see any downsides to this, especially since there is no action associated with the egress rule.

Also, in your case, you are rejecting an entire rule even if one domain conflicts. If you follow the merge procedure (which is a simple change in the function above), then you would be basically collecting a set of virtualhost:port entries along with the respective setting for use_egress_proxy. This can then be used to generate the virtual host blocks, etc.

What do you think @kyessenov ?

Reply:
I would go with the simple solution. In the documentation, state that egress rules cannot share the same domain with the same port. Period. Then we just remove conflicting rules altogether, instead of implementing complex merge semantics. This way the users are protected from breaking their Istio deployment, but their conflicting rules will be rejected and will appear in the pilot logs as such. Once rejected, it will be more clearly noticeable to the users - they will get 404 for the domains in the rule. If we start partially allowing conflicting rules, it will be harder to notice and to understand what was merged and why.

Reply:
We can go even further and for simplicity require that no two rules can share the same domain, even for different ports. This way one domain can appear in one rule only, and if there is a need to add a port, the containing rule will have to be updated.

Reply:
@frankbu what is your opinion on this?
Egress rule is basically a white list.

Reply:
Since there is only one attribute (use_egress_proxy) that may conflict, it seems that we could get away without validating this for now and simply merge. I think the "safe" merge rule is that if any rule requests use_egress_proxy == true for a domain then use_egress_proxy will be true (OR semantics) for that domain. It seems that use_egress_proxy == false is just an optimization.

If we do that for now, we can always add validation in the future, if this causes problems.

Reply:
@frankbu Please note that also protocols on the port can conflict. For example, you can have three rules on the same domain, that would claim that some port on this domain supports protocols `http`, `http2` or `https`. The question is which protocol to select while merging these three rules into one.

1. Should `https` win over `http` and `http2` ? The outbound cluster will have SSL context so Envoy will perform TLS origination.
2. Should `http2` win over `http`? The outbound cluster will have `http2` feature defined.

Reply:
Lets merge this as it is today.. Once we update egress rule to contain a single domain, this should go away.